ksql-cli/src/main/java/io/confluent/ksql/cli/Cli.java

```diff
@@ -510,4 +466,105 @@ private void setRequestPipelining(final boolean newSetting) {
       requestPipelining = newSetting;
     }
   }
+
+  private class QueryStreamSubscriber extends BaseSubscriber<StreamedRow> {
+
+    private final CompletableFuture<Void> future;
+    private boolean closed;
+    private long rowsRead;
+
+    QueryStreamSubscriber(final Context context, final CompletableFuture<Void> future) {
+      super(context);
+      this.future = Objects.requireNonNull(future);
+    }
+
+    @Override
+    protected void afterSubscribe(final Subscription subscription) {
+      makeRequest(1);
+    }
+
+    @Override
+    protected synchronized void handleValue(final StreamedRow row) {
+      if (closed) {
+        return;
+      }
+      terminal.printStreamedRow(row);
+      terminal.flush();
+      if (row.isTerminal()) {
+        future.complete(null);
+        close();
+        return;
+      }
+      if (row.getRow().isPresent()) {
+        rowsRead++;
+        if (streamedQueryRowLimit != null && streamedQueryRowLimit == rowsRead) {
+          future.complete(null);
+          close();
+          return;
+        }
+      }
+      makeRequest(1);
+    }
+
+    @Override
+    protected void handleComplete() {
+      future.complete(null);
+    }
+
+    @Override
+    protected void handleError(final Throwable t) {
+      future.completeExceptionally(t);
+    }
+
+    synchronized void close() {
+      closed = true;
+      context.runOnContext(v -> cancel());
+    }
+  }
+
+  private class PrintTopicSubscriber extends BaseSubscriber<String> {
+
+    private final CompletableFuture<Void> future;
+    private boolean closed;
+
+    PrintTopicSubscriber(final Context context, final CompletableFuture<Void> future) {
+      super(context);
+      this.future = Objects.requireNonNull(future);
+    }
+
+    @Override
+    protected void afterSubscribe(final Subscription subscription) {
+      makeRequest(1);
+    }
+
+    @Override
+    protected synchronized void handleValue(final String line) {
+      if (closed) {
+        return;
+      }
+      if (line.isEmpty()) {
+        close();
```

Comment:
Closing on newlines does not seem valid, since the TopicStreamWriter may insert them: https://github.com/confluentinc/ksql/blob/6bcbd51d9c057593b950d8be4c1a4e708c36567a/ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/TopicStreamWriter.java#L107

I see you've got handling for similar new lines from the QueryStreamWriter, and a test for it in KsqlClientTest. Can we add the analogous for print topic handling as well?

Reply:
Ack, good catch.